File: packages/app-interaction-shell/src/param-gathering-fsm.test.ts
import { describe, expect, it } from "vitest";
import { LengthMm, PointMm } from "@toporoom/domain-floorplan";
import { ParamGatheringFSM } from "./param-gathering-fsm.js";
import { WallDrawTool } from "./wall-draw-tool.js";

describe("ParamGatheringFSM / WallDrawTool", () => {
  it("collects start → end → thickness then yields an AddWall draft", () => {
    const fsm = ParamGatheringFSM.forTool(new WallDrawTool());
    expect(fsm.state).toBe("awaiting_start");

    fsm.provide("start", PointMm.of(0, 0));
    expect(fsm.state).toBe("awaiting_end");

    fsm.provide("end", PointMm.of(4000, 0));
    expect(fsm.state).toBe("awaiting_thickness");

    const draft = fsm.provide("thickness", LengthMm.of(200));
    expect(fsm.state).toBe("complete");
    expect(draft).toEqual({
      type: "AddWall",
      start: { x: 0, y: 0 },
      end: { x: 4000, y: 0 },
      thicknessMm: 200,
    });
  });

  it("rejects parameters that are out of order", () => {
    const fsm = ParamGatheringFSM.forTool(new WallDrawTool());
    expect(() => fsm.provide("end", PointMm.of(1, 0))).toThrow(/start/i);
  });
});

File: core/tests/param_gathering_fsm_test.cpp
#include <gtest/gtest.h>

#include "toporoom/app/param_gathering_fsm.hpp"
#include "toporoom/app/wall_draw_tool.hpp"

using toporoom::app::WallDrawTool;

TEST(ParamGatheringFSM, CollectsStartEndThickness) {
  WallDrawTool tool;
  auto fsm = tool.create_fsm();
  EXPECT_EQ(fsm.state(), "awaiting_start");
  fsm.provide("start", "0,0");
  EXPECT_EQ(fsm.state(), "awaiting_end");
  fsm.provide("end", "4000,0");
  EXPECT_EQ(fsm.state(), "awaiting_thickness");
  fsm.provide("thickness", "200");
  EXPECT_EQ(fsm.state(), "complete");
  const auto draft = WallDrawTool::complete(fsm.params());
  EXPECT_EQ(draft.type, "AddWall");
  EXPECT_EQ(draft.start_x, 0);
  EXPECT_EQ(draft.end_x, 4000);
  EXPECT_EQ(draft.thickness_mm, 200);
}

TEST(ParamGatheringFSM, RejectsOutOfOrder) {
  auto fsm = WallDrawTool().create_fsm();
  EXPECT_THROW(fsm.provide("end", "1,0"), std::runtime_error);
}
